site/website/blog/2020-10-06-enabling-full-text-search.md

```diff
@@ -0,0 +1,492 @@
+
+The next step in the pipeline is to extract the keywords from the visit
+description. We will use Rapid Automatic Keyword Extraction (RAKE)
+algorithm, the implementation we use was originally published on
+[Github](https://github.com/Linguistic/rake). The implementation is not
+important for the demonstration, you could use any other Java library,
+call 3rd party service, e.g. via [grpc](/docs/how-tos/grpc) or use our
+[python integration](/docs/tutorials/python).
+
+```java
+// Create factory for keyword service
+ServiceFactory<?, Rake> keywordService = ServiceFactories.sharedService((context) -> new Rake("en"));
+
+p.readFrom(mysqlSource)
+ .withoutTimestamps()
+ .mapUsingService(keywordService, PetClinicIndexJob::enrichWithKeywords)
+```
+
+The method `enrichWithKeywords` simply passes the description field from
+the change record to the keyword extraction service and sets the results
+into the keywords field of the Visit instance.
+
+It is a common pattern to enrich an item in a pipeline with more
+information, Hazelcast Jet works with POJOs in the pipeline, so you need
+to somehow compose the original item and the enriching information. One
+option is to use tuples. This works for a small number of fields. Other
+option would be to use a domain-specific object holding all the
+information - e.g. the `keywords` field. If you are on Java 14 or newer
+you might want to give the [Records preview
+feature](https://openjdk.java.net/jeps/359) a try.
+
+### Joining
+
+Change data capture sends changes to the database as individual records
+(represented by
+[ChangeRecord](https://jet-start.sh/javadoc/4.2/com/hazelcast/jet/cdc/ChangeRecord.html).
+Each record represents a changed row in a database table. If a single
+row is all that you need then you are good to go. You can see such
+example in our [Evergreen Cache blog
+post](/blog/2020/07/16/designing-evergreen-cache-cdc). But having a
+single row to work with is not always the case - often you need to
+aggregate either multiple records from the same table or join records
+from different tables into a single aggregate record.
+
+In our case, we need to join 3 tables into a single document, which we
+then index into Elasticsearch. To perform the joining we will use
+`mapStateful` step. This can be performed either globally or partitioned
+on a grouping key. If all your records share a common key it is always a
+good idea to do the grouping by the key, because the mapping state is
+then evenly distributed across all nodes and the mapping operation
+parallelized.
+
+Our visit record contains the pet id but not the owner id, in such case,
+so we don’t have a single grouping key for all records. There are two options: 
+* perform the stateful mapping globally, which has the advantage
+of having a single state - the joining logic is simpler and the pipeline
+more straightforward. The obvious disadvantage is that such a solution
+is not scalable because the state may not fit onto a single member or it
```

Comment:
```suggestion
is not scalable because the state may not fit onto a single member and it
```

Reply:
It's one or the other, or both at the same time. Not sure what is correct in English.

Reply:
I guess it should say "because the state is on a single member and the computation is single-threaded."